### ISSUE TITLE
Incorrect padding while using NetworkWithInputEncoding

### ISSUE THREAD

**Issue by WeiPhil** [NVlabs/tiny-cuda-nn#145]
Hi,
First of all, thank you for this amazing library! 

Recently, I found a network configuration that seems to pose problems with input alignments. My network triggers a `tiny-cuda-nn/src/fully_fused_mlp.cu:592 check failed in_width % 16 == 0`, it uses a composite input encoding where the first 3 dimensions use an Identity encoding and the 3 subsequent encodings use the hashgrid encoding with default parameters. The output dimension is one-dimensional. While I would expect the `NetworkWithInputEncoding` to automatically manage the output padding the network indicates `m_encoding->padded_output_width() = 61` and thus triggers the assert.
Is this some sort of odd incorrectly handled case or am I missing something. Note that swapping the first 3 with the next 3 dimensions doesn't trigger the issue. 

Best,

**Comment by Tom94** [comment]
Hi, could you attach the full config of the composite encoding and its constituent encodings so I can try reproducing on my end? Thanks!

**Comment by WeiPhil** [comment]
Of course, sorry.
```
  tcnn::json config = {
      {"loss", {{"otype", "L2"}}},
      {"optimizer",
       {{"otype", "Adam"},
        {"learning_rate", 0.01},}},
      {"encoding",
       {{"otype", "Composite"},
        {"nested",
         {{
              {"n_dims_to_encode", 3},
              {"otype", "Identity"},
          },
          {
              {"n_dims_to_encode", 3},
              {"otype", "HashGrid"},
              {"n_levels", 4},
              {"n_features_per_level", 2},
              {"log2_hashmap_size", 19},
              {"base_resolution", 32},
              {"per_level_scale", 2.0},
              {"interpolation",  "Linear"}
          }}}}},
      {"network",
       {
           {"otype", "FullyFusedMLP"},
           {"n_neurons", 64},
           {"n_hidden_layers", 2},
           {"activation", "ReLU"},
           {"output_activation", "Sigmoid"},
       }},
  };

  tcnn::json encoding_opts = config.value("encoding", tcnn::json::object());
  tcnn::json loss_opts = config.value("loss", tcnn::json::object());
  tcnn::json optimizer_opts = config.value("optimizer", tcnn::json::object());
  tcnn::json network_opts = config.value("network", tcnn::json::object());

  auto loss = std::shared_ptr<tcnn::Loss<tcnn::network_precision_t>>{
      tcnn::create_loss<tcnn::network_precision_t>(loss_opts)};
  auto optimizer = std::shared_ptr<tcnn::Optimizer<tcnn::network_precision_t>>{
      tcnn::create_optimizer<tcnn::network_precision_t>(optimizer_opts)};
  auto network = std::make_shared<tcnn::NetworkWithInputEncoding<tcnn::network_precision_t>>(
      visibility_n_input_dims, visibility_n_output_dims, encoding_opts, network_opts);

  auto trainer = std::make_shared<
      tcnn::Trainer<float, tcnn::network_precision_t, tcnn::network_precision_t>>(
      network, optimizer, loss);
```

**Comment by Tom94** [comment]
Thank you very much, this helps. It looks like the issue is somewhat deep and requires me to re-write (and first re-think) the way padding and alignment interplay with each other in order to generate as optimal code as reasonable.

I probably won't get to this until next week, but what I _can_ already say is that, speed-wise, it'll almost always be best to put the encodings with fewest output dimensions last. So I'd recommend sticking with the reverted order that you've already discovered, even when the bug is fixed.

Cheers!

**Comment by WeiPhil** [comment]
Thank you so much for the rapid answer, I'll stick to my temporary solution in the meantime then :)

**Comment by Tom94** [comment]
Should be fixed now. Please feel free to re-open the issue if you encounter more breakage. Thanks again for the detailed report.
